File: examples/ag-ui/servers/php/lib/Agui.php
<?php

declare(strict_types=1);

const DEFAULT_OPENAI_MODEL = 'gpt-4o';
const DEFAULT_ANTHROPIC_MODEL = 'claude-sonnet-4-6';

function props_from_input(array $input): array
{
    if (!empty($input['forwardedProps']) && is_array($input['forwardedProps'])) {
        return $input['forwardedProps'];
    }

    if (!empty($input['data']) && is_array($input['data'])) {
        return $input['data'];
    }

    return [];
}

function provider_from_input(array $input): array
{
    $props = props_from_input($input);
    $provider = is_string($props['provider'] ?? null) && $props['provider'] !== ''
        ? $props['provider']
        : 'openai';
    $model = is_string($props['model'] ?? null) ? $props['model'] : '';

    if ($provider === 'anthropic') {
        return [
            'provider' => 'anthropic',
            'model' => $model !== '' ? $model : DEFAULT_ANTHROPIC_MODEL,
        ];
    }

    return [
        'provider' => 'openai',
        'model' => $model !== '' ? $model : DEFAULT_OPENAI_MODEL,
    ];
}

function sse_event(array $payload): string
{
    return 'data: ' . json_encode($payload, JSON_UNESCAPED_UNICODE) . "\n\n";
}

function sse_done(): string
{
    return "data: [DONE]\n\n";
}

function run_started(string $threadId, string $runId): string
{
    return sse_event([
        'type' => 'RUN_STARTED',
        'threadId' => $threadId,
        'runId' => $runId,
    ]);
}

function text_message_start(string $messageId): string
{
    return sse_event([
        'type' => 'TEXT_MESSAGE_START',
        'messageId' => $messageId,
        'role' => 'assistant',
    ]);
}

function text_message_content(string $messageId, string $delta): string
{
    return sse_event([
        'type' => 'TEXT_MESSAGE_CONTENT',
        'messageId' => $messageId,
        'delta' => $delta,
    ]);
}

function text_message_end(string $messageId): string
{
    return sse_event([
        'type' => 'TEXT_MESSAGE_END',
        'messageId' => $messageId,
    ]);
}

function run_finished(string $threadId, string $runId): string
{
    return sse_event([
        'type' => 'RUN_FINISHED',
        'threadId' => $threadId,
        'runId' => $runId,
        'finishReason' => 'stop',
    ]);
}

function run_error(string $threadId, string $runId, string $message): string
{
    return sse_event([
        'type' => 'RUN_ERROR',
        'threadId' => $threadId,
        'runId' => $runId,
        'error' => ['message' => $message],
    ]);
}

function begin_sse(): void
{
    header('Content-Type: text/event-stream');
    header('Cache-Control: no-cache');
    header('Connection: keep-alive');
    header('Access-Control-Allow-Origin: *');
    http_response_code(200);
}

function write_sse(string $frame): void
{
    echo $frame;
    if (ob_get_level() > 0) {
        ob_flush();
    }
    flush();
}

function cors_preflight(): void
{
    header('Access-Control-Allow-Origin: *');
    header('Access-Control-Allow-Methods: POST, OPTIONS');
    header('Access-Control-Allow-Headers: Content-Type');
    http_response_code(204);
}

function stream_text(
    string $threadId,
    string $runId,
    callable $stream
): void {
    begin_sse();
    $messageId = 'msg-' . $runId;

    write_sse(run_started($threadId, $runId));
    write_sse(text_message_start($messageId));

    $emit = static function (string $delta) use ($messageId): void {
        if ($delta === '') {
            return;
        }
        write_sse(text_message_content($messageId, $delta));
    };

    try {
        $stream($emit);
        write_sse(text_message_end($messageId));
        write_sse(run_finished($threadId, $runId));
        write_sse(sse_done());
    } catch (Throwable $error) {
        error_log('[php] chat error: ' . $error->getMessage());
        write_sse(run_error($threadId, $runId, $error->getMessage()));
        write_sse(sse_done());
    }
}

function stream_error_only(string $threadId, string $runId, string $message): void
{
    begin_sse();
    write_sse(run_started($threadId, $runId));
    write_sse(run_error($threadId, $runId, $message));
    write_sse(sse_done());
}

File: examples/ag-ui/servers/go/agui.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
)

const listenAddr = ":8001"

const (
	defaultOpenAIModel    = "gpt-4o"
	defaultAnthropicModel = "claude-sonnet-4-6"
)

type runAgentInput struct {
	ThreadID       string            `json:"threadId"`
	RunID          string            `json:"runId"`
	Messages       []incomingMessage `json:"messages"`
	ForwardedProps map[string]any    `json:"forwardedProps"`
	Data           map[string]any    `json:"data"`
}

type incomingMessage struct {
	Role    string          `json:"role"`
	Content json.RawMessage `json:"content"`
	Parts   []messagePart   `json:"parts"`
}

type messagePart struct {
	Type    string `json:"type"`
	Content string `json:"content"`
}

type chatMessage struct {
	Role    string
	Content string
}

type providerConfig struct {
	Provider string
	Model    string
}

type sseStream struct {
	w          http.ResponseWriter
	flusher    http.Flusher
	threadID   string
	runID      string
	messageID  string
	started    bool
	textOpened bool
}

func propsFromInput(input runAgentInput) map[string]any {
	if len(input.ForwardedProps) > 0 {
		return input.ForwardedProps
	}
	return input.Data
}

func providerFromInput(input runAgentInput) providerConfig {
	props := propsFromInput(input)
	provider := "openai"
	model := ""

	if raw, ok := props["provider"].(string); ok && raw != "" {
		provider = raw
	}
	if raw, ok := props["model"].(string); ok && raw != "" {
		model = raw
	}

	switch provider {
	case "anthropic":
		if model == "" {
			model = defaultAnthropicModel
		}
	default:
		provider = "openai"
		if model == "" {
			model = defaultOpenAIModel
		}
	}

	return providerConfig{Provider: provider, Model: model}
}

func beginSSE(w http.ResponseWriter) (*sseStream, bool) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		return nil, false
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.WriteHeader(http.StatusOK)

	return &sseStream{w: w, flusher: flusher}, true
}

func (s *sseStream) initRun(threadID, runID string) {
	s.threadID = threadID
	s.runID = runID
	s.messageID = fmt.Sprintf("msg-%s", runID)
}

func (s *sseStream) writeEvent(payload map[string]any) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	_, err = fmt.Fprintf(s.w, "data: %s\n\n", data)
	if err != nil {
		return err
	}
	s.flusher.Flush()
	return nil
}

func (s *sseStream) writeDone() {
	_, _ = fmt.Fprint(s.w, "data: [DONE]\n\n")
	s.flusher.Flush()
}

func (s *sseStream) startRun() error {
	if s.started {
		return nil
	}
	s.started = true
	return s.writeEvent(map[string]any{
		"type":     "RUN_STARTED",
		"threadId": s.threadID,
		"runId":    s.runID,
	})
}

func (s *sseStream) startText() error {
	if s.textOpened {
		return nil
	}
	s.textOpened = true
	return s.writeEvent(map[string]any{
		"type":      "TEXT_MESSAGE_START",
		"messageId": s.messageID,
		"role":      "assistant",
	})
}

func (s *sseStream) writeDelta(delta string) error {
	if delta == "" {
		return nil
	}
	return s.writeEvent(map[string]any{
		"type":      "TEXT_MESSAGE_CONTENT",
		"messageId": s.messageID,
		"delta":     delta,
	})
}

func (s *sseStream) finishSuccess() error {
	if s.textOpened {
		if err := s.writeEvent(map[string]any{
			"type":      "TEXT_MESSAGE_END",
			"messageId": s.messageID,
		}); err != nil {
			return err
		}
	}
	return s.writeEvent(map[string]any{
		"type":         "RUN_FINISHED",
		"threadId":     s.threadID,
		"runId":        s.runID,
		"finishReason": "stop",
	})
}

func (s *sseStream) finishError(message string) error {
	if !s.started {
		if err := s.startRun(); err != nil {
			return err
		}
	}
	return s.writeEvent(map[string]any{
		"type":     "RUN_ERROR",
		"threadId": s.threadID,
		"runId":    s.runID,
		"error": map[string]any{
			"message": message,
		},
	})
}

func (s *sseStream) streamText(stream func(emit func(string) error) error) error {
	if err := s.startRun(); err != nil {
		return err
	}
	if err := s.startText(); err != nil {
		return err
	}
	if err := stream(s.writeDelta); err != nil {
		return err
	}
	if err := s.finishSuccess(); err != nil {
		return err
	}
	s.writeDone()
	return nil
}
